File: basic/regular_api.js
/**
 * RoraOS Regular Chat API - JavaScript Example
 * =============================================
 *
 * Example of using the Regular Chat API for chat completion.
 * This API is compatible with OpenAI format.
 *
 * Usage in Node.js:
 *   node regular_api.js
 *
 * Usage in Browser:
 *   Import the required functions
 */

// Configuration
const API_URL = "https://labs.roraos.com/api/v1/chat";
const API_KEY = "your-api-key-here"; // Get from dashboard

/**
 * Send chat completion request to API
 *
 * @param {Array} messages - Array of messages [{role, content}]
 * @param {Object} options - Additional options
 * @param {boolean} options.stream - Use streaming
 * @param {number} options.temperature - Creativity (0-2)
 * @param {number} options.max_tokens - Maximum tokens
 * @returns {Promise} Response from API
 */
async function chatCompletion(messages, options = {}) {
  const { stream = false, temperature = 0.7, max_tokens = 2000 } = options;

  const payload = {
    messages,
    model: "gpt-4o", // Model displayed, backend uses available model
    stream,
    temperature,
    max_tokens,
  };

  if (stream) {
    return chatCompletionStream(payload);
  }

  const response = await fetch(API_URL, {
    method: "POST",
    headers: {
      Authorization: `Bearer ${API_KEY}`,
      "Content-Type": "application/json",
    },
    body: JSON.stringify(payload),
  });

  if (!response.ok) {
    const error = await response.json();
    throw new Error(error.error || "API request failed");
  }

  return response.json();
}

/**
 * Handle streaming response
 */
async function chatCompletionStream(payload) {
  const response = await fetch(API_URL, {
    method: "POST",
    headers: {
      Authorization: `Bearer ${API_KEY}`,
      "Content-Type": "application/json",
    },
    body: JSON.stringify(payload),
  });

  if (!response.ok) {
    const error = await response.json();
    throw new Error(error.error || "API request failed");
  }

  const reader = response.body.getReader();
  const decoder = new TextDecoder();
  let fullContent = "";

  while (true) {
    const { done, value } = await reader.read();
    if (done) break;

    const chunk = decoder.decode(value);
    const lines = chunk.split("\n");

    for (const line of lines) {
      if (line.startsWith("data: ")) {
        const data = line.slice(6);
        if (data === "[DONE]") break;

        try {
          const parsed = JSON.parse(data);
          const content =
            parsed.content ||
            parsed.choices?.[0]?.delta?.content ||
            "";
          if (content) {
            fullContent += content;
            process.stdout.write(content);
          }
        } catch {
          // Skip invalid JSON
        }
      }
    }
  }

  console.log(); // New line
  return fullContent;
}

// ============ Usage Examples ============

async function main() {
  // Example 1: Simple chat
  console.log("=".repeat(50));
  console.log("Example 1: Simple Chat");
  console.log("=".repeat(50));

  try {
    const messages = [{ role: "user", content: "Hello! Who are you?" }];
    const result = await chatCompletion(messages);
    console.log(`Response: ${result.choices[0].message.content}`);
  } catch (error) {
    console.log(`Error: ${error.message}`);
  }

  // Example 2: Chat with context (multi-turn)
  console.log("\n" + "=".repeat(50));
  console.log("Example 2: Multi-turn Chat");
  console.log("=".repeat(50));

  try {
    const messages = [
      {
        role: "system",
        content: "You are a helpful assistant.",
      },
      { role: "user", content: "What is JavaScript?" },
      {
        role: "assistant",
        content: "JavaScript is a programming language for web development.",
      },
      { role: "user", content: "Give me a simple code example." },
    ];

    const result = await chatCompletion(messages, { temperature: 0.5 });
    console.log(`Response: ${result.choices[0].message.content}`);
  } catch (error) {
    console.log(`Error: ${error.message}`);
  }

  // Example 3: Streaming response
  console.log("\n" + "=".repeat(50));
  console.log("Example 3: Streaming Response");
  console.log("=".repeat(50));

  try {
    const messages = [
      { role: "user", content: "Explain AI in 3 sentences." },
    ];
    process.stdout.write("Response: ");
    await chatCompletion(messages, { stream: true });
  } catch (error) {
    console.log(`Error: ${error.message}`);
  }
}

// Run examples
main().catch(console.error);

// Export for use as module
if (typeof module !== "undefined") {
  module.exports = { chatCompletion, chatCompletionStream };
}

File: basic/regular_api.py
"""
RoraOS Regular Chat API - Python Example
=========================================

Example of using the Regular Chat API for chat completion.
This API is compatible with OpenAI format.

Installation:
    pip install requests

Usage:
    python regular_api.py
"""

import requests
import json

# Configuration
API_URL = "https://labs.roraos.com/api/v1/chat"
API_KEY = "your-api-key-here"  # Get from dashboard


def chat_completion(messages: list, stream: bool = False, temperature: float = 0.7, max_tokens: int = 2000):
    """
    Send chat completion request to API.

    Args:
        messages: List of messages with format [{"role": "user", "content": "Hello"}]
        stream: Use streaming response
        temperature: Creativity (0.0 - 2.0)
        max_tokens: Maximum response tokens

    Returns:
        Response from API
    """
    headers = {
        "Authorization": f"Bearer {API_KEY}",
        "Content-Type": "application/json"
    }

    payload = {
        "messages": messages,
        "model": "gpt-4o",  # Model displayed, backend uses available model
        "stream": stream,
        "temperature": temperature,
        "max_tokens": max_tokens
    }

    if stream:
        return chat_completion_stream(headers, payload)

    response = requests.post(API_URL, headers=headers, json=payload)
    response.raise_for_status()
    return response.json()


def chat_completion_stream(headers: dict, payload: dict):
    """Handle streaming response."""
    response = requests.post(
        API_URL,
        headers=headers,
        json=payload,
        stream=True
    )
    response.raise_for_status()

    full_content = ""
    for line in response.iter_lines():
        if line:
            line = line.decode('utf-8')
            if line.startswith('data: '):
                data = line[6:]
                if data == '[DONE]':
                    break
                try:
                    chunk = json.loads(data)
                    content = chunk.get('choices', [{}])[0].get('delta', {}).get('content', '')
                    if content:
                        full_content += content
                        print(content, end='', flush=True)
                except json.JSONDecodeError:
                    pass

    print()  # New line after streaming
    return full_content


# ============ Usage Examples ============

if __name__ == "__main__":
    # Example 1: Simple chat
    print("=" * 50)
    print("Example 1: Simple Chat")
    print("=" * 50)

    messages = [
        {"role": "user", "content": "Hello! Who are you?"}
    ]

    try:
        result = chat_completion(messages)
        print(f"Response: {result['choices'][0]['message']['content']}")
    except requests.exceptions.RequestException as e:
        print(f"Error: {e}")

    # Example 2: Chat with context (multi-turn)
    print("\n" + "=" * 50)
    print("Example 2: Multi-turn Chat")
    print("=" * 50)

    messages = [
        {"role": "system", "content": "You are a helpful assistant."},
        {"role": "user", "content": "What is Python?"},
        {"role": "assistant", "content": "Python is a popular high-level programming language."},
        {"role": "user", "content": "Give me a simple code example."}
    ]

    try:
        result = chat_completion(messages, temperature=0.5)
        print(f"Response: {result['choices'][0]['message']['content']}")
    except requests.exceptions.RequestException as e:
        print(f"Error: {e}")

    # Example 3: Streaming response
    print("\n" + "=" * 50)
    print("Example 3: Streaming Response")
    print("=" * 50)

    messages = [
        {"role": "user", "content": "Explain AI in 3 sentences."}
    ]

    try:
        print("Response: ", end='')
        chat_completion(messages, stream=True)
    except requests.exceptions.RequestException as e:
        print(f"Error: {e}")
